lib/puppet_forge_server/backends/proxy.rb

```diff
@@ -59,6 +63,27 @@ def upload(file_data)
     protected
     attr_reader :log
 
+    def get_non_mutable(relative_url)
+      file_name = relative_url.split('/').last
+      target_file = File.join(@cache_dir, file_name[0].downcase, file_name)
+      path = Dir["#{@cache_dir}/**/#{file_name}"].first
+      unless File.exist?("#{path}")
+        buffer = get(relative_url)
+        File.open(target_file, 'wb') do |file|
+          file.write(buffer)
+          @log.debug("Saved #{buffer} bytes in filesystem cache for url: #{relative_url}, target file: #{target_file}")
+        end
+        path = target_file
+      else
+        @log.info("Filesystem cache HIT for url: #{relative_url}")
```

Comment:
What do you mean?

This is main speedup factor - to cache non mutable file (metadata about specific version) on HDD